backend/de.metas.manufacturing/src/main/java/de/metas/distribution/ddordercandidate/material_dispo/DDOrderCandidateAdvisedEventCreator.java

```diff
@@ -0,0 +1,193 @@
+import de.metas.material.event.supplyrequired.SupplyRequiredDecreasedEvent;
+import de.metas.material.planning.MaterialPlanningContext;
+import de.metas.material.planning.ProductPlanning;
+import de.metas.material.planning.ddordercandidate.DDOrderCandidateDataFactory;
+import de.metas.material.planning.ddordercandidate.DDOrderCandidateDemandMatcher;
+import de.metas.material.planning.event.SupplyRequiredAdvisor;
+import de.metas.quantity.Quantity;
+import de.metas.quantity.Quantitys;
+import de.metas.uom.UomId;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+@RequiredArgsConstructor
+public class DDOrderCandidateAdvisedEventCreator implements SupplyRequiredAdvisor
+{
+	@NonNull private final DDOrderCandidateDemandMatcher demandMatcher;
+	@NonNull private final DDOrderCandidateDataFactory ddOrderCandidateDataFactory;
+	@NonNull private final DDOrderCandidateService ddOrderCandidateService;
+	@NonNull private final CandidateRepositoryWriteService candidateRepositoryWriteService;
+	@NonNull private final CandidateRepositoryRetrieval candidateRepositoryRetrieval;
+
+	public List<DDOrderCandidateAdvisedEvent> createAdvisedEvents(
+			@NonNull final SupplyRequiredDescriptor supplyRequiredDescriptor,
+			@NonNull final MaterialPlanningContext context)
+	{
+		if (!demandMatcher.matches(context))
+		{
+			return ImmutableList.of();
+		}
+
+		final ProductPlanning productPlanningData = context.getProductPlanning();
+
+		return ddOrderCandidateDataFactory.create(supplyRequiredDescriptor, context)
+				.stream()
+				.map(ddOrderCandidate -> toDDOrderCandidateAdvisedEvent(ddOrderCandidate, supplyRequiredDescriptor, productPlanningData))
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private static DDOrderCandidateAdvisedEvent toDDOrderCandidateAdvisedEvent(
+			@NonNull final DDOrderCandidateData ddOrderCandidate,
+			@NonNull final SupplyRequiredDescriptor supplyRequiredDescriptor,
+			@NonNull final ProductPlanning productPlanningData)
+	{
+		return DDOrderCandidateAdvisedEvent.builder()
+				.eventDescriptor(supplyRequiredDescriptor.newEventDescriptor())
+				.supplyRequiredDescriptor(supplyRequiredDescriptor)
+				.ddOrderCandidate(ddOrderCandidate)
+				.advisedToCreateDDOrder(productPlanningData.isCreatePlan())
+				.pickIfFeasible(productPlanningData.isPickDirectlyIfFeasible())
+				.build();
+	}
+
+	@Override
+	public BigDecimal handleQuantityDecrease(final @NonNull SupplyRequiredDecreasedEvent event,
+											 @NonNull final BigDecimal qtyToDistribute)
+	{
+		final Set<DDOrderCandidateId> candidateIds = getDDOrderCandidateIds(event);
+		if (qtyToDistribute.signum() <= 0 || candidateIds.isEmpty())
```

Comment:
minor: for optimization purposes i think u can move the if qtyToDistribute.signum() <= 0 part before calling getDDOrderCandidateIds.